cel/pkg/reconciler/cel/celrun.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2021 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run"
+	"knative.dev/pkg/logging"
+
+	v1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	reconciler "knative.dev/pkg/reconciler"
+)
+
+// newReconciledNormal makes a new reconciler event with event type Normal, and reason RunReconciled.
+func newReconciledNormal(namespace, name string) reconciler.Event {
+	return reconciler.NewEvent(v1.EventTypeNormal, "RunReconciled", "Run reconciled: \"%s/%s\"", namespace, name)
+}
+
+// Reconciler implements controller.Reconciler for Run resources.
+type Reconciler struct {
+}
+
+// Check that our Reconciler implements Interface
+var _ run.Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *v1alpha1.Run) reconciler.Event {
+	logger := logging.FromContext(ctx)
+	logger.Infof("Reconciling Run %s/%s", o.Namespace, o.Name)
+
+	// If the Run has not started, initialize the Condition and set the start time.
+	if !o.HasStarted() {
+		logger.Infof("Starting new Run %s/%s", o.Namespace, o.Name)
+		o.Status.InitializeConditions()
+		// In case node time was not synchronized, when controller has been scheduled to other nodes.
+		if o.Status.StartTime.Sub(o.CreationTimestamp.Time) < 0 {
+			logger.Warnf("Run %s/%s createTimestamp %s is after the Run started %s", o.Namespace, o.Name, o.CreationTimestamp, o.Status.StartTime)
+			o.Status.StartTime = &o.CreationTimestamp
+		}
+	}
+
+	if o.IsDone() {
+		logger.Infof("Run %s/%s is done", o.Namespace, o.Name)
+		return nil
+	}
+
+	// Create a program environment configured with the standard library of CEL functions and macros
+	env, err := cel.NewEnv(cel.Declarations())
+	if err != nil {
+		logger.Errorf("cel.NewEnv: %v", err)
+		return err
+	}
+
+	// Iterate through the CEL expressions
+	for _, param := range o.Spec.Params {
```

Comment:
I think we want to pull out the param named `expression`, fail if it isn't found, and only evaluate that param value. This should also fail if any other params are defined, since that's probably a typo/bug.

I don't think we want/need to support multiple expressions in one run, since it's a bit unclear how it should work, and it's probably clearer (though much more verbose) to just have multiple CEL-typed Runs if you want that.

Reply:
> since it's a bit unclear how it should work, and it's probably clearer (though much more verbose) to just have multiple CEL-typed Runs if you want that.

the way I have it implemented now is that the `Run` will evaluate each `Parameter` and produce a `Result` with the corresponding name (i.e. parameter name == result name), did that to reduce the verbosity and evaluate multiple expressions in one pod when appropriate, what do you think of that?

Reply:
Yeah, I think I'm coming around to it. It is better than having N single-expression Runs.

One possible confusing thing is that if a CEL Run has three expression params and one is an invalid expression, will it fail but still produce results for the other two expressions? Will it still evaluate params defined after the invalid expression? There are some options:

1. Fail and don't report any results if any expression param is invalid.
2. (this is what this PR does) Evaluate each expression in the order they're defined in `params`, fail when one is invalid and only report results for previous valid evaluations.
3. Evaluate each expression, fail when one is invalid but keep evaluating params and report results for any that are valid expressions. You can report either the first, or _all_, invalid expressions in `status`.

Either way I don't think it should block this PR, it'd be better to get usage and feedback.

Reply:
As a user I would expect it to fail if any param expression is invalid. I think I assume a full validation before it runs

Reply:
I've added an earlier validation to check that that at least one expression is provided and that the expression types are valid

To be consistent with `Tasks` which don't produce `Results` when they fail, I've modified so that we go with option 1 where we don't report any evaluation result if any expression is invalid or has any syntax/evaluation error -- and I agree that we can revisit those options based on user feedback